### DISCUSSION TITLE
How to move js script from head to body end?

### DISCUSSION THREAD

**Discussion by amit-yadav-30082990** [vercel/next.js#42886] [Help]
After building nextjs application, the html page contains js scripts in head tag, how we can move them to the body end?

**Comment by icyJoseph** [comment]
Is there any script in particular you want to place after the body? 

`defer` is an attribute added to scripts to make sure the DOM is loaded by the time the scripts run. Most scripts injected by Next.js have `defer` set though.

Is there any reason why you need to do this?

## Solution

OP needs to put a preload tag on the `_document` tag, which would tell the browser to load this asset before the deferred scripts. No further changes are needed, more here https://github.com/vercel/next.js/discussions/42886#discussioncomment-4155003

**Reply by amit-yadav-30082990** [reply]
Need to move all defer js as they are blocking browser and hold to load further assets till the js get loaded ([wait for an available TCP socket](https://stackoverflow.com/questions/34043505/chrome-devtools-what-does-queueing-means-in-the-timing-tab)), that impact page LCP. Specially in case of mweb / slow internet connection
[WPT Report](https://www.webpagetest.org/result/221114_AiDcZ3_77T/1/details/#waterfall_view_step1)

**Reply by icyJoseph** [reply]
> Scripts with the defer attribute will execute in the order in which they appear in the document.
> 
> This attribute allows the elimination of parser-blocking JavaScript where the browser would have to load and evaluate scripts before continuing to parse. async has a similar effect in this case.

According to the MDN docs https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script, that shouldn't really be the case, and what would be your strategy with placing them at the body tag end, IIRC, defer is slightly better because the scripts are run, after entire DOM is parsed and loaded, though the event for that is not called just yet.

In this case LCP is affected because the PNG is loaded as a resource invoked in a CSS stylesheet, which btw is also preloaded, so the browser is giving the image as higher priority as it can. Unfortunately there's only so much we can do when the device power and connection is as low as in that report, and that PNG weights `117.1kB`.

> Downloading `react-dom` on 3G takes ~840ms.

Could you explain your reasoning further?

**Reply by amit-yadav-30082990** [reply]
Thanks for the observation on it, but e.g. I have 10 JS files which are deferred in head tag, and two images are there to impact LCP are in body tag, and because of slow internet / latency in response in JS file browser will wait to get free TCP socket to load further images / resources. 
Ofcoruse in case of deferred it will not block the rendering but it blocks TCP sockets.

**Reply by icyJoseph** [reply]
And so you want to move them to the end of the body to place them at the end of the deferred queue, which would allow you to flush the images, correct?

These 10 JS files, are they under your control? If so, you could place them after the `NextScript`, in `_document`.

**Reply by amit-yadav-30082990** [reply]
These JS file are created and handled by nextjs builder only. So I don't have control on it. there are 16 js file on [next.js](https://nextjs.org/) home page, can we control on it?

<img width="775" alt="image" src="https://user-images.githubusercontent.com/112924105/202130515-afeb6d4a-f0c9-49c3-96f0-14eb8f67a6d2.png">

**Reply by icyJoseph** [reply]
The amount of JS files of course varies from project to project, but at the very last there'll be a webpack, framework, main and pages specific chunks, plus the manifest files.

I recall someone who has using a tool that did something slightly similar. In this exact comment, https://github.com/vercel/next.js/discussions/41299#discussioncomment-3840077 they talk about a plugin that rewrote the script tags, but it kept them in place though!

I don't think there's a simple way, if any at all, to do this as part of the toolchain, but with some shell scripting I guess one could target all `.html` files and move the scripts further down, sounds super complex and brittle though.

## An alternative

Wouldn't it be easier to move requests for your images "further up on the DOM"? After all, deferred resources are loaded as they are queued, so technically, all you need is your images to come before anything else. 

So you could place something like:

```html
  <link rel="preload" href="my-image.png" as="image" />
```

I think if you put this in `_document` Head, then it'll be placed before the script tags, which means the browser goes to fetch it ASAP. More on the MDN site, https://developer.mozilla.org/en-US/docs/Web/HTML/Link_types/preload.

I did a small experiment, and you can see that with the preload directive FP, FCP and LCP, happen almost next to each other:

![Screenshot 2022-11-16 at 10 32 53](https://user-images.githubusercontent.com/21013447/202143473-d058597e-6e0f-4b52-8d10-3f2e6d360251.png)

If I remove the preload, notice the LCP gets displaced very far to the right:

![Screenshot 2022-11-16 at 10 37 32](https://user-images.githubusercontent.com/21013447/202144602-a2212439-6f94-45f2-baae-c3ad7abfc53a.png)

> Note, because I use next/image, the preload actually looks like:

```html
<link
      rel="preload"
      href="/_next/image?url=%2Fmy-image.jpg&w=1080&q=75"
      as="image"
    />
```

And the entire `_document`:

```tsx
import { Html, Head, Main, NextScript } from "next/document";

export default function Document() {
  return (
    <Html>
      <Head>
        <link
          rel="preload"
          href="/_next/image?url=%2Fmy-image.jpg&w=1080&q=75"
          as="image"
        />
      </Head>
      <body>
        <Main />
        <NextScript />
      </body>
    </Html>
  );
}
```

**Reply by amit-yadav-30082990** [reply]
Yes we can go with preload thing but when we are using such a huge framework, I think there must be some next config to move it down to body end. I hope this config will be there in upcoming release :)

**Reply by icyJoseph** [reply]
I wouldn't call Next.js a huge framework though, rather it is a framework that closes on its responsibilities while allow the developer to interact and configure the HTML results. 

In this case, there's a escape hatch at the Head tag to preload content, and it is such a natural HTML pattern that even a library such as SWR recommends it.

**Reply by Lolhp0** [reply]
im also waiting for something like this I rather script tags to be in the body rather than in the head. if someone has a solution please do tell me. btw im aware of the using a normal script tag to move them into the body but if theres a way to use nextjs config or something please tell me
